File: lib/twine/formatters/android.rb
# encoding: utf-8
require 'cgi'
require 'rexml/document'

module Twine
  module Formatters
    class Android < Abstract
      include Twine::Placeholders

      SUPPORTS_PLURAL = true
      LANG_CODES = Hash[
        'zh' => 'zh-Hans',
        'zh-CN' => 'zh-Hans',
        'zh-HK' => 'zh-Hant',
        # See https://developer.android.com/reference/java/util/Locale#legacy-language-codes
        'iw' => 'he',
        'in' => 'id',
        'ji' => 'yi'
      ]

      def format_name
        'android'
      end

      def extension
        '.xml'
      end

      def can_handle_directory?(path)
        Dir.entries(path).any? { |item| /^values.*$/.match(item) }
      end

      def default_file_name
        'strings.xml'
      end

      def determine_language_given_path(path)
        path_arr = path.split(File::SEPARATOR)
        path_arr.each do |segment|
          if segment == 'values'
            return @twine_file.language_codes[0]
          else
            # The language is defined by a two-letter ISO 639-1 language code, optionally followed by a two letter ISO 3166-1-alpha-2 region code (preceded by lowercase "r").
            # see http://developer.android.com/guide/topics/resources/providing-resources.html#AlternativeResources
            match = /^values-([a-z]{2}(-r[a-z]{2})?)$/i.match(segment)

            if match
              lang = match[1].sub('-r', '-')
              return LANG_CODES.fetch(lang, lang)
            end
          end
        end

        return super
      end

      def output_path_for_language(lang)
        if lang == @twine_file.language_codes[0]
          "values"
        else
          "values-#{lang}".gsub(/-(\p{Lu})/, '-r\1')
        end
      end

      def set_translation_for_key(key, lang, value)
        value = CGI.unescapeHTML(value)
        value.gsub!('\\\'', '\'')
        value.gsub!('\\"', '"')
        value = convert_placeholders_from_android_to_twine(value)
        value.gsub!('\@', '@')
        value.gsub!(/(\\u0020)*|(\\u0020)*\z/) { |spaces| ' ' * (spaces.length / 6) }
        super(key, lang, value)
      end

      def read(io, lang)
        document = REXML::Document.new io, :compress_whitespace => %w{ string }
        document.context[:attribute_quote] = :quote
        comment = nil
        document.root.children.each do |child|
          if child.is_a? REXML::Comment
            content = child.string.strip
            content.gsub!(/[\s]+/, ' ')
            comment = content if content.length > 0 and not content.start_with?("SECTION:")
          elsif child.is_a? REXML::Element
            next unless child.name == 'string'

            key = child.attributes['name']

            content = child.children.map(&:to_s).join
            set_translation_for_key(key, lang, content)
            set_comment_for_key(key, comment) if comment

            comment = nil
          end 
        end
      end

      def format_header(lang)
        "<?xml version=\"1.0\" encoding=\"utf-8\"?>\n"
      end

      def format_sections(twine_file, lang)
        result = '<resources>'
        
        result += super + "\n"

        result += "</resources>\n"
      end

      def format_section_header(section)
        "#{space(4)}<!-- SECTION: #{section.name} -->"
      end

      def format_comment(definition, lang)
        "#{space(4)}<!-- #{definition.comment.gsub('--', '—')} -->\n" if definition.comment
      end

      def key_value_pattern
        "#{space(4)}<string name=\"%{key}\">%{value}</string>"
      end

      def format_plural_keys(key, plural_hash)
        result = "#{space(4)}<plurals name=\"#{key}\">\n"
        result += plural_hash.map{|quantity,value| "#{space(8)}<item quantity=\"#{quantity}\">#{escape_value(value)}</item>"}.join("\n")
        result += "\n#{space(4)}</plurals>"
      end

      def space(level)
        ' ' * level
      end

      def gsub_unless(text, pattern, replacement)
        text.gsub(pattern) do |match|
          match_start_position = Regexp.last_match.offset(0)[0]
          yield(text[0, match_start_position]) ? match : replacement
        end
      end

      # http://developer.android.com/guide/topics/resources/string-resource.html#FormattingAndStyling
      def escape_value(value)
        inside_cdata = /<\!\[CDATA\[((?!\]\]>).)*$/              # opening CDATA tag ('<![CDATA[') not followed by a closing tag (']]>')
        inside_opening_tag = /<(a|font|span|p)\s?((?!>).)*$/     # tag start ('<a ', '<font ', '<span ' or '<p ') not followed by a '>'

        # escape double and single quotes and & signs
        value = gsub_unless(value, '"', '\\"') { |substring| substring =~ inside_cdata || substring =~ inside_opening_tag }
        value = gsub_unless(value, "'", "\\'") { |substring| substring =~ inside_cdata }
        value = gsub_unless(value, /&/, '&amp;') { |substring| substring =~ inside_cdata || substring =~ inside_opening_tag }

        # if `value` contains a placeholder, escape all angle brackets
        # if not, escape opening angle brackes unless it's a supported styling tag
        # https://github.com/scelis/twine/issues/212
        # https://stackoverflow.com/questions/3235131/#18199543
        if number_of_twine_placeholders(value) > 0 or @options[:escape_all_tags]
          # matches all `<` but <![CDATA
          angle_bracket = /<(?!(\/?(\!\[CDATA)))/
        else
          # matches all '<' but <b>, <em>, <i>, <cite>, <dfn>, <big>, <small>, <font>, <tt>, <s>,
          # <strike>, <del>, <u>, <super>, <sub>, <ul>, <li>, <br>, <div>, <span>, <p>, <a>
          # and <![CDATA
          angle_bracket = /<(?!(\/?(b|em|i|cite|dfn|big|small|font|tt|s|strike|del|u|super|sub|ul|li|br|div|span|p|a|\!\[CDATA)))/
        end
        value = gsub_unless(value, angle_bracket, '&lt;') { |substring| substring =~ inside_cdata }
        value = gsub_unless(value, '\n', "\n\\n") { |substring| substring =~ inside_cdata }

        # escape non resource identifier @ signs (http://developer.android.com/guide/topics/resources/accessing-resources.html#ResourcesFromXml)
        resource_identifier_regex = /@(?!([a-z\.]+:)?[a-z+]+\/[a-zA-Z_]+)/   # @[<package_name>:]<resource_type>/<resource_name>
        value.gsub(resource_identifier_regex, '\@')
      end

      # see http://developer.android.com/guide/topics/resources/string-resource.html#FormattingAndStyling
      # however unescaped HTML markup like in "Welcome to <b>Android</b>!" is stripped when retrieved with getString() (http://stackoverflow.com/questions/9891996/)
      def format_value(value)
        value = value.dup

        # convert placeholders (e.g. %@ -> %s)
        value = convert_placeholders_from_twine_to_android(value)

        # capture xliff tags and replace them with a placeholder
        xliff_tags = []
        value.gsub! /<xliff:g.+?<\/xliff:g>/ do
          xliff_tags << $&
          'TWINE_XLIFF_TAG_PLACEHOLDER'
        end

        # escape everything outside xliff tags
        value = escape_value(value)

        # put xliff tags back into place
        xliff_tags.each do |xliff_tag|
          # escape content of xliff tags
          xliff_tag.gsub! /(<xliff:g.*?>)(.*)(<\/xliff:g>)/ do "#{$1}#{escape_value($2)}#{$3}" end
          value.sub! 'TWINE_XLIFF_TAG_PLACEHOLDER', xliff_tag
        end
        
        # replace beginning and end spaces with \u0020. Otherwise Android strips them.
        value.gsub(/\A *| *\z/) { |spaces| '\u0020' * spaces.length }
      end

    end
  end
end

Twine::Formatters.formatters << Twine::Formatters::Android.new

File: python_twine/twine/formatters/android.py
"""
Android XML strings formatter.
"""

import re
import html
from typing import Dict, Optional, TextIO
from xml.etree import ElementTree as ET
from xml.etree.ElementTree import Element

from twine.formatters import AbstractFormatter
from twine.formatters.tools import replace_with_filter
from twine.placeholders import (
    convert_placeholders_from_android_to_twine,
    convert_placeholders_from_twine_to_android,
    number_of_twine_placeholders,
)


def inner_xml(node:Element) -> str:
    # Get inner XML (text + nested elements)
    # Start with the text before any child element
    value = node.text or ""

    # Add each child element's XML and tail
    for subelement in node:
        value += ET.tostring(subelement, encoding="unicode", method="html")
    return value

class AndroidFormatter(AbstractFormatter):
    """Formatter for Android XML string resources."""

    SUPPORTS_PLURAL = True

    # Language code mappings for Android
    ANDROID_TO_TWINE_LANG_CODES = {
        "zh": "zh-Hans",
        "zh-TW": "zh-Hant",
        "zh-CN": "zh-Hans",
        "zh-HK": "zh-Hant",
        # Legacy language codes
        "iw": "he",
        "in": "id",
        "ji": "yi",
    }

    TWINE_TO_ANDROID_LANG_CODES = {
        "zh-Hans": "zh",
        "zh-Hant": "zh-TW",
        "he": "iw",
        "id": "in",
        "yi": "ji",
    }

    def format_name(self) -> str:
        return "android"

    def extension(self) -> str:
        return ".xml"

    def can_handle_directory(self, path: str) -> bool:
        """Check if directory contains Android values folders."""
        import os

        try:
            entries = os.listdir(path)
            return any(item.startswith("values") for item in entries)
        except (OSError, IOError):
            return False

    def default_file_name(self) -> str:
        return "strings.xml"

    def determine_language_given_path(self, path: str) -> Optional[str]:
        """Extract language from Android path like values-es-rMX."""
        from pathlib import Path

        path_parts = Path(path).parts

        for segment in path_parts:
            # Default values folder is developer language
            if segment == "values":
                if self.twine_file.language_codes:
                    return self.twine_file.language_codes[0]
                return None

            # values-{lang} or values-{lang}-r{region}
            match = re.match(
                r"^values-([a-z]{2,3}(-r[a-z]{2,4})?)$", segment, re.IGNORECASE
            )
            if match:
                lang = match.group(1).replace("-r", "-")
                return self.ANDROID_TO_TWINE_LANG_CODES.get(lang, lang)

        return super().determine_language_given_path(path)

    def output_path_for_language(self, lang: str) -> str:
        """Get Android values folder name for language."""
        if self.twine_file.language_codes and lang == self.twine_file.language_codes[0]:
            return "values"
        else:
            lang = self.TWINE_TO_ANDROID_LANG_CODES.get(lang, lang)
            # Convert en-US to values-en-rUS
            result = f"values-{lang}"
            result = re.sub(r"-([A-Z])", r"-r\1", result)
            return result

    def read(self, io: TextIO, lang: str):
        """Read Android XML strings file."""
        content = io.read()

        # Parse with comment support
        try:
            parser = ET.XMLParser(target=ET.TreeBuilder(insert_comments=True))
            root = ET.fromstring(content, parser)
        except ET.ParseError as e:
            from twine import TwineError

            raise TwineError(f"Failed to parse XML: {e}")

        comment = None
        current_section = None

        for child in root:
            # Handle comments (they have a callable tag function)
            if callable(child.tag):
                content_text = child.text.strip() if child.text else ""
                content_text = re.sub(r"\s+", " ", content_text)
                if content_text:
                    if content_text.startswith("SECTION:"):
                        current_section = content_text[8:].strip()
                    else:
                        comment = content_text

            # Handle string elements
            elif child.tag == "string":
                key = child.get("name")
                if not key:
                    continue

                value = self.unescape_value(inner_xml(child))
                self.set_translation_for_key(key, lang, value, current_section)

                if comment:
                    self.set_comment_for_key(key, comment)
                    comment = None

            # Handle plural strings elements:
            #  <plurals name="bookmarks_places">
            #    <item quantity="one">%d bookmark</item>
            #    <item quantity="other">%d bookmarks</item>
            #  </plurals>
            elif child.tag == "plurals":
                key = child.get("name")
                if not key:
                    continue

                plural_values = {}
                for subelement in child:
                    if subelement.tag == "item":
                        quantity = subelement.get("quantity")
                        if not quantity:
                            continue
                        plural_values[quantity] = self.unescape_value(inner_xml(subelement))

                if plural_values:
                    self.set_translation_for_key_plural(key, lang, plural_values, current_section)

                if comment:
                    self.set_comment_for_key(key, comment)
                    comment = None

    def format_header(self, lang: str) -> str:
        return '<?xml version="1.0" encoding="utf-8"?>'

    def format_sections(self, twine_file, lang: str) -> str:
        result = "<resources>"

        sections_content = super().format_sections(twine_file, lang)
        if sections_content:
            result += "\n" + sections_content + "\n"

        result += "</resources>\n"
        return result

    def format_section_header(self, section) -> str:
        return f"    <!-- SECTION: {section.name} -->"

    def format_comment(self, definition, lang: str) -> Optional[str]:
        if definition.comment:
            # Replace -- with em dash to avoid XML comment issues
            comment = definition.comment.replace("--", "—")
            return f"    <!-- {comment} -->\n"
        return None

    def key_value_pattern(self) -> str:
        return '    <string name="%(key)s">%(value)s</string>'

    def format_plural_keys(self, key: str, plural_hash: Dict[str, str]) -> str:
        """Format Android plurals."""
        result = f'    <plurals name="{key}">\n'

        items = []
        for quantity, value in plural_hash.items():
            escaped_value = self.escape_value(value)
            items.append(f'        <item quantity="{quantity}">{escaped_value}</item>')

        result += "\n".join(items)
        result += "\n    </plurals>"
        return result

    @staticmethod
    def unescape_value(value: str) -> str:
        """ Unescape HTML entities """
        value = html.unescape(value)

        # Unescape Android escapes
        value = value.replace("\\'", "'")
        value = value.replace('\\"', '"')

        # Convert placeholders from Android to Twine
        value = convert_placeholders_from_android_to_twine(value)

        # Unescape @ signs
        value = value.replace("\\@", "@")

        # Unescape \n
        value = value.replace("\n\\n", "\n")

        # Convert \u0020 space escapes
        def replace_spaces(match):
            spaces = match.group(0)
            return " " * (len(spaces) // 6)

        return re.sub(r"(\\u0020)+", replace_spaces, value)

    def escape_value(self, value: str) -> str:
        """
        Escape value for Android XML.

        http://developer.android.com/guide/topics/resources/string-resource.html#FormattingAndStyling
        """

        # Check if inside CDATA or opening tag
        def inside_cdata(text: str, pos: int) -> bool:
            before = text[:pos]
            # Check if we're inside a CDATA that hasn't closed
            return (
                "<![CDATA[" in before
                and "]]>" not in before[before.rfind("<![CDATA[") :]
            )

        def inside_opening_tag(text: str, pos: int) -> bool:
            before = text[:pos]
            # Check if we're inside an opening tag
            match = re.search(r"<(a|font|span|p)\s+[^>]*$", before)
            return match is not None

        # Escape double quotes (unless in CDATA or opening tag)
        result = replace_with_filter(value, '"', '\\"',
            lambda i: not (inside_cdata(value, i) or inside_opening_tag(value, i))
        )

        # Escape single quotes (unless in CDATA)
        result = replace_with_filter(result, "'", "\\'",
            lambda i: not inside_cdata(result, i)
        )

        # Escape ampersands (unless in CDATA or opening tag)
        result = replace_with_filter(result, "&", "&amp;",
            lambda i: not (inside_cdata(result, i) or inside_opening_tag(result, i))
        )

        # Escape angle brackets based on placeholder presence
        has_placeholders = number_of_twine_placeholders(value) > 0

        if has_placeholders or self.options.get("escape_all_tags"):
            # Escape all < except <![CDATA
            angle_bracket_regex = re.compile(r"<(?!(\/?(\!\[CDATA)))")
        else:
            # Escape < except supported tags
            angle_bracket_regex = re.compile(
                r"<(?!(\/?(b|em|i|cite|dfn|big|small|font|tt|s|strike|del|u|super|sub|ul|li|br|div|span|p|a|\!\[CDATA))\b)"
            )

        def is_non_tag(result:str, i:int):
            if inside_cdata(result, i):
                return False
            # Check if this '<' isn't a known tag
            remaining = result[i:]
            return angle_bracket_regex.match(remaining) #is not None

        result = replace_with_filter(result, "<", "&lt;",
            lambda i: is_non_tag(result, i)
        )

        # Escape newlines (unless in CDATA)
        result = replace_with_filter(result, "\\n", "\n\\n",
            lambda i: not inside_cdata(result, i)
        )

        # Escape @ signs that aren't resource identifiers
        resource_identifier_regex = re.compile(r"@(?!([a-z\.]+:)?[a-z+]+\/[a-zA-Z_]+)")
        result = resource_identifier_regex.sub(r"\\@", result)

        return result

    def format_value(self, value: str) -> str:
        """Format value for Android output."""
        # Convert placeholders
        value = convert_placeholders_from_twine_to_android(value)

        # Handle xliff:g tags
        xliff_tags = []

        def save_xliff(match):
            xliff_tags.append(match.group(0))
            return "TWINE_XLIFF_TAG_PLACEHOLDER"

        value = re.sub(r"<xliff:g.+?</xliff:g>", save_xliff, value)

        # Escape everything outside xliff tags
        value = self.escape_value(value)

        # Restore xliff tags with escaped content
        for xliff_tag in xliff_tags:
            # Escape content inside xliff tags
            escaped_xliff = re.sub(
                r"(<xliff:g.*?>)(.*)(</xliff:g>)",
                lambda m: m.group(1) + self.escape_value(m.group(2)) + m.group(3),
                xliff_tag,
            )
            value = value.replace("TWINE_XLIFF_TAG_PLACEHOLDER", escaped_xliff, 1)

        # Replace beginning and end spaces with \u0020
        value = re.sub(r"^ +| +$", lambda m: "\\u0020" * len(m.group(0)), value)

        return value
